controllers/replication.storage/volumegroupreplication_controller.go

```diff
@@ -33,21 +56,533 @@ type VolumeGroupReplicationReconciler struct {
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+		CreateFunc: func(e event.CreateEvent) bool {
+			return false
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			return false
+		},
+	}
+
+	enqueueVGRRequest := handler.EnqueueRequestsFromMapFunc(
+		func(context context.Context, obj client.Object) []reconcile.Request {
+			// Get the VolumeGroupReplication name,namespace
+			var vgrName, vgrNamespace string
+			objAnnotations := obj.GetAnnotations()
+			for k, v := range objAnnotations {
+				if k == "replication.storage.openshift.io/volumegroupref" {
+					vgrName = strings.Split(v, "/")[0]
+					vgrNamespace = strings.Split(v, "/")[1]
+					break
+				}
+			}
+
+			// Skip reconcile if the triggering resource is not a sub-resource of VGR
+			if vgrName == "" || vgrNamespace == "" {
+				return []reconcile.Request{}
+			}
+
+			// Check if the resource is present in the cluster
+			vgrObj := &replicationv1alpha1.VolumeGroupReplication{}
+			logger := log.FromContext(context)
+			err := r.Client.Get(context, types.NamespacedName{Name: vgrName, Namespace: vgrNamespace}, vgrObj)
+			if err != nil {
+				logger.Error(err, "Unable to get VolumeGroupReplication resource")
+				return []reconcile.Request{}
+			}
+
+			// Return name and namespace of the VoleumGroupReplication resource
+			return []reconcile.Request{
+				{
+					NamespacedName: types.NamespacedName{
+						Namespace: vgrObj.Namespace,
+						Name:      vgrObj.Name,
+					},
+				},
+			}
+		},
+	)
+
+	enqueueVGRForPVCRequest := handler.EnqueueRequestsFromMapFunc(
+		func(context context.Context, obj client.Object) []reconcile.Request {
+			// Check if the PVC has any labels defined
+			objLabels := obj.GetLabels()
+			if len(objLabels) == 0 {
+				return []reconcile.Request{}
+			}
+
+			// Check if the resource is present in the cluster
+			vgrObjsList := &replicationv1alpha1.VolumeGroupReplicationList{}
+			logger := log.FromContext(context)
+			err := r.Client.List(context, vgrObjsList)
+			if err != nil {
+				logger.Error(err, "failed to list VolumeGroupReplication instances")
+				return []reconcile.Request{}
+			}
+
+			// Check if the pvc labels match any VGRs based on selectors present in it's annotation
+			for _, vgr := range vgrObjsList.Items {
+				if vgr.Annotations != nil && vgr.Annotations["pvcSelector"] != "" {
+					labelSelector, err := labels.Parse(vgr.Annotations["pvcSelector"])
+					if err != nil {
+						logger.Error(err, "failed to parse selector from VolumeGroupReplication's annotation")
+						return []reconcile.Request{}
+					}
+					objLabelsSet := labels.Set(objLabels)
+					if labelSelector.Matches(objLabelsSet) {
+						return []reconcile.Request{
+							{
+								NamespacedName: types.NamespacedName{
+									Namespace: vgr.Namespace,
+									Name:      vgr.Name,
+								},
+							},
+						}
+					}
+				}
+			}
```

Comment:
there should be a controller runtime util function for it check something like https://github.com/kubernetes-sigs/controller-runtime/blob/a0c9fd9d3f310f48155ce985366b21914675fbea/pkg/client/example_test.go#L265-L294

Reply:
Well, I will just update the code to fetch the VGR name from the annotation, since now we are adding the owner annotation to PVC.

Reply:
The thing is we can't use a helper function like the one you mentioned is because the selector is not a single field that we can compare or so, and to match labels we need to create a selector requirement as done in the getPVCFromDataSource function.